### PR TITLE
Add docker pull and push tasks

### PR DESCRIPTION
**Pull request by barbelity**

### CODE REVIEW

File: src/main/resources/atlassian-plugin.xml [hunk 1]
```diff
@@ -235,6 +235,21 @@
         <exporter class="org.jfrog.bamboo.exporter.ArtifactoryTaskExporter"/>
     </taskType>
 
+    <!--Docker task-->
+    <taskType key="artifactoryDockerTask" name="Artifactory Docker"
+              class="org.jfrog.bamboo.task.ArtifactoryDockerTask">
+        <description>Push docker images to Artifactory.</description>
+        <configuration class="org.jfrog.bamboo.configuration.ArtifactoryDockerConfiguration"/>
```
Comment (yahavi) [inline src/main/resources/atlassian-plugin.xml:242]: Please add the runtimeTaskDataProvider
```suggestion
        <configuration class="org.jfrog.bamboo.configuration.ArtifactoryDockerConfiguration"/>
        <runtimeTaskDataProvider class="org.jfrog.bamboo.security.provider.SharedCredentialsDataProvider"/>

```

File: src/main/resources/templates/plugins/task/editArtifactoryDockerBuilder.ftl [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+                resolveErrorDiv.innerHTML = errorMessage;
+                resolveErrorDiv.style.display = '';
+                resolveRepoSelect.innerHTML = '';
+            }
+        });
+    }
+
+    function displayPublishingDockerArtifactoryConfigs(serverId) {
+        var configDiv = document.getElementById('dockerArtifactoryPublishingConfigDiv');
+        var credentialsUserName = configDiv.getElementsByTagName('input')[1].value;
```
Comment (yahavi) [inline src/main/resources/templates/plugins/task/editArtifactoryDockerBuilder.ftl:129]:
```suggestion
        var credentialsUserName = configDiv.getElementsByTagName('input')[2].value;
```

File: src/main/resources/templates/plugins/task/editArtifactoryDockerBuilder.ftl [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        var credentialsPassword = configDiv.getElementsByTagName('input')[2].value;
```
Comment (yahavi) [inline src/main/resources/templates/plugins/task/editArtifactoryDockerBuilder.ftl:130]:
```suggestion
        var credentialsPassword = configDiv.getElementsByTagName('input')[3].value;
```

**Review by yahavi — CHANGES_REQUESTED** [review]
I just reviewed the shared credentials parts. Please see my comments.

File: src/main/resources/atlassian-plugin.xml (hunk 1 above)
Comment (barbelity) [reply src/main/resources/atlassian-plugin.xml:242]: Added.

File: src/main/java/org/jfrog/bamboo/configuration/ArtifactoryDockerConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    @Override
+    public boolean taskProducesTestResults(@NotNull TaskDefinition taskDefinition) {
+        return false;
+    }
+
+    @Override
+    public void validate(@NotNull ActionParametersMap params, @NotNull ErrorCollection errorCollection) {
+        String commandChoiceKey = NpmBuildContext.COMMAND_CHOICE;
```
Comment (yahavi) [inline src/main/java/org/jfrog/bamboo/configuration/ArtifactoryDockerConfiguration.java:116]:
```suggestion
        String commandChoiceKey = COMMAND_CHOICE;
```

File: src/main/java/org/jfrog/bamboo/context/DockerBuildContext.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    public static Set<String> getFieldsToCopy() {
+        return Sets.newHashSet(COMMAND_CHOICE, COMMAND_PULL, COMMAND_PUSH, DOCKER_PUSH_SERVER_ID, DOCKER_PULL_SERVER_ID,
+                DOCKER_PULL_REPO, DOCKER_PUSH_REPO, DOCKER_PULL_USERNAME, DOCKER_PULL_PASSWORD, DOCKER_PUSH_USERNAME,
+                HOST, IMAGE_NAME, DOCKER_PUSH_PASSWORD, INCLUDE_ENV_VARS_PARAM, ENV_VARS_EXCLUDE_PATTERNS,
+                ENV_VARS_INCLUDE_PATTERNS, CAPTURE_BUILD_INFO, BUILD_NAME, BUILD_NUMBER, RESOLVER_OVERRIDE_CREDENTIALS_CHOICE,
+                RESOLVER_SHARED_CREDENTIALS, DEPLOYER_OVERRIDE_CREDENTIALS_CHOICE, DEPLOYER_SHARED_CREDENTIALS);
+    }
+
+    public boolean isDockerCommandPull() {
+        return (ArtifactoryDockerConfiguration.CFG_DOCKER_COMMAND_PULL.equals(env.get(COMMAND_CHOICE)));
```
Comment (yahavi) [inline src/main/java/org/jfrog/bamboo/context/DockerBuildContext.java:44]: Redundant parenthesis
```suggestion
        return ArtifactoryDockerConfiguration.CFG_DOCKER_COMMAND_PULL.equals(env.get(COMMAND_CHOICE));
```

**Review by yahavi — CHANGES_REQUESTED** [review]
Very good!
Please consider to delete `viewArtifactoryDeploymentAction.ftl` as well.

File: src/main/java/org/jfrog/bamboo/configuration/ArtifactoryDockerConfiguration.java (hunk 4 above)
Comment (barbelity) [reply src/main/java/org/jfrog/bamboo/configuration/ArtifactoryDockerConfiguration.java:116]: Fixed.

File: src/main/java/org/jfrog/bamboo/context/DockerBuildContext.java (hunk 5 above)
Comment (barbelity) [reply src/main/java/org/jfrog/bamboo/context/DockerBuildContext.java:44]: Fixed.

**Comment by barbelity** [comment]
> Very good!
> Please consider to delete `viewArtifactoryDeploymentAction.ftl` as well.

You are correct, removed `viewArtifactoryDeploymentAction.ftl`.
Thanks 👍

**Pull request merged**